src/genkey/clu_genkey.c

```diff
@@ -462,122 +499,152 @@ int wolfCLU_genKey_RSA(WC_RNG* rng, char* fName, int directive, int fmt, int
+                    break;
+                }
+                FALL_THROUGH;
+            case PUB_ONLY_FILE:
+                /* add on the final part of the file name ".pub" */
+                if (ret == WOLFCLU_SUCCESS) {
+                    XMEMCPY(fOutNameBuf + fNameSz, fExtPub, fExtSz);
+                    WOLFCLU_LOG(WOLFCLU_L0, "fOutNameBuf = %s", fOutNameBuf);
 
-            ret = (int)XFWRITE(derBuf, 1, derBufSz, file);
-            if (ret <= 0) {
-                XFCLOSE(file);
-                XFREE(fOutNameBuf, HEAP_HINT, DYNAMIC_TYPE_TMP_BUFFER);
-                XFREE(derBuf, HEAP_HINT, DYNAMIC_TYPE_TMP_BUFFER);
-                return OUTPUT_FILE_ERROR;
-            }
-            XFCLOSE(file);
-            break;
-        default:
-            WOLFCLU_LOG(WOLFCLU_L0, "Invalid directive");
-            XFREE(fOutNameBuf, HEAP_HINT, DYNAMIC_TYPE_TMP_BUFFER);
-            XFREE(derBuf, HEAP_HINT, DYNAMIC_TYPE_TMP_BUFFER);
-            return BAD_FUNC_ARG;
-    }
+                    derBufSz = wc_RsaKeyToPublicDer(&key, derBuf,
+                            (word32)maxDerBufSz);
+                    if (derBufSz < 0) {
+                        ret = derBufSz;
+                    }
+                    outBuf   = derBuf;
+                    outBufSz = derBufSz;
+                }
 
-    XFREE(derBuf, HEAP_HINT, DYNAMIC_TYPE_TMP_BUFFER);
-    XFREE(fOutNameBuf, HEAP_HINT, DYNAMIC_TYPE_TMP_BUFFER);
-    wc_FreeRsaKey(&key);
+                /* check if should convert to PEM format */
+                if (ret == WOLFCLU_SUCCESS && fmt == PEM_FORM) {
+                    pemBufSz = wolfCLU_KeyDerToPem(derBuf, derBufSz, &pemBuf,
+                            PUBLICKEY_TYPE, DYNAMIC_TYPE_PUBLIC_KEY);
+                    if (pemBufSz <= 0 || pemBuf == NULL) {
+                        ret =  WOLFCLU_FAILURE;
+                    }
+                    outBuf   = pemBuf;
+                    outBufSz = pemBufSz;
+                }
 
-    if (ret > 0) {
-        /* ret > 0 indicates a successful file write, set to zero for return */
-        ret = WOLFCLU_SUCCESS;
+                if (ret == WOLFCLU_SUCCESS) {
+                    file = XFOPEN(fOutNameBuf, "wb");
+                    if (file == XBADFILE) {
+                        ret = OUTPUT_FILE_ERROR;
+                    }
+                }
+
+                if (ret == WOLFCLU_SUCCESS) {
+                    if ((int)XFWRITE(outBuf, 1, outBufSz, file) <= 0) {
+                        ret = OUTPUT_FILE_ERROR;
+                    }
+                    XFCLOSE(file);
+                }
+
+                if (pemBuf != NULL) {
+                    XFREE(pemBuf, HEAP_HINT, DYNAMIC_TYPE_PUBLIC_KEY);
+                }
+
+                break;
+            default:
+                WOLFCLU_LOG(WOLFCLU_L0, "Invalid directive");
+                ret = BAD_FUNC_ARG;
+        }
+    }
+
+    if (derBuf != NULL) {
+        XFREE(derBuf, HEAP_HINT, DYNAMIC_TYPE_TMP_BUFFER);
```

Comment:
Should we 0 the DER buffer, too, in addition to the PEM one you're already 0-ing?